File: python/1144. Decrease Elements To Make Array Zigzag.py
class Solution(object):
    def movesToMakeZigzag(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        odd = even = 0
        for i in range(n):
            if i&1: # i-1, i+1是偶数
                d1 = nums[i] - nums[i-1] + 1 if nums[i] >= nums[i-1] else 0
                d2 = nums[i] - nums[i+1] + 1 if i < n-1 and nums[i] >= nums[i+1] else 0
                odd += max(d1, d2)
            else:
                d1 = nums[i] - nums[i-1] + 1 if i > 0 and nums[i] >= nums[i-1] else 0
                d2 = nums[i] - nums[i+1] + 1 if i < n-1 and nums[i] >= nums[i+1] else 0
                even += max(d1, d2)
        return min(odd, even)

# 思路：下标分成奇数或者偶数的情况分别计算

File: go/src/leetcode/1144/1144. Decrease Elements To Make Array Zigzag.go
package main

import "fmt"

/*
*
https://leetcode.cn/problems/decrease-elements-to-make-array-zigzag/
*/
func movesToMakeZigzag(nums []int) int {
	n := len(nums)
	var odd, even int
	for i := range nums {
		if i&1 == 1 { // i-1, i+1 is even
			var l, r int
			if i-1 >= 0 && nums[i] >= nums[i-1] {
				l = nums[i] - nums[i-1] + 1
			}
			if i+1 < n && nums[i] >= nums[i+1] {
				r = nums[i] - nums[i+1] + 1
			}
			odd += max(l, r)
		} else { // i-1, i+1 is odd
			var l, r int
			if i-1 >= 0 && nums[i] >= nums[i-1] {
				l = nums[i] - nums[i-1] + 1
			}
			if i+1 < n && nums[i] >= nums[i+1] {
				r = nums[i] - nums[i+1] + 1
			}
			even += max(l, r)
		}
	}
	return min(odd, even)
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	fmt.Println(movesToMakeZigzag([]int{9, 6, 1, 6, 2}))
}
